integration-tests/java/src/integration-test/java/com/stellarstation/api/test/satellite/SatelliteStreamerTest.java

```diff
@@ -55,67 +56,57 @@ void setUp() {
-                  ByteString data = response.getReceiveTelemetryResponse().getTelemetry().getData();
-                  if (data.size() > 2) {
-                    // The second last byte of the telemetry indicates the current state of the
-                    // fake satellite used in the test. The value is either of 0 or 1.
-                    int state = data.byteAt(data.size() - 2);
-                    if (initialState < 0) {
-                      initialState = state;
-                    } else {
-                      commandTestFuture.set(initialState == 1 - state);
-                    }
-
-                    // Send a command to toggle the state.
-                    SendSatelliteCommandsRequest commandsRequest =
-                        SendSatelliteCommandsRequest.newBuilder()
-                            .addCommand(ByteString.copyFrom(new byte[] {0x01, 0x01}))
-                            .build();
-                    Futures.getUnchecked(requestObserverFuture)
-                        .onNext(
-                            SatelliteStreamRequest.newBuilder()
-                                .setSatelliteId(SATELLITE_ID)
-                                .setSendSatelliteCommandsRequest(commandsRequest)
-                                .build());
-                  }
-                } else {
-                  GroundStationState gsState =
-                      response.getStreamEvent().getPlanMonitoringEvent().getGroundStationState();
-                  AntennaState antennaState = gsState.getAntenna();
-
-                  azimuthCommand.add(antennaState.getAzimuth().getCommand());
-                  azimuthMeasured.add(antennaState.getAzimuth().getMeasured());
-                  elevationCommand.add(antennaState.getElevation().getCommand());
-                  elevationMeasured.add(antennaState.getElevation().getMeasured());
-                }
-              }
-
-              @Override
-              public void onError(Throwable t) {
-                commandTestFuture.setException(t);
-              }
-
-              @Override
-              public void onCompleted() {}
-            });
+    class TelemetryAndCommandTestStreamObserver implements StreamObserver<SatelliteStreamResponse> {
+      private List<Integer> safeModeStates = new ArrayList();
+
+      @Override
+      public void onNext(SatelliteStreamResponse response) {
+        if (response.hasReceiveTelemetryResponse()) {
+          ByteString data = response.getReceiveTelemetryResponse().getTelemetry().getData();
+          if (data.size() > 2) {
+            // The second last byte of the telemetry indicates the current state of the
+            // fake satellite used in the test. The value is either of 0 or 1.
+            int state = data.byteAt(data.size() - 2);
+            safeModeStates.add(state);
+
+            // Send a command to toggle the state.
+            SendSatelliteCommandsRequest commandsRequest =
```

Comment:
I was expecting some kind of condition to stop us toggling forever. Maybe we should only send this toggle command when safeModeStates.size() == 1? I'm assuming that when you toggle, you will get a future 'onNext' which will then cause us to toggle again...and repeats forever.

Reply:
The most flaky part of the current fake satellite is accepting command and toggling status. I want to have a concrete code that can stop the code even when it is broken.